geo-traits/src/point.rs

```diff
@@ -0,0 +1,177 @@
+use geo_types::{Coord, CoordNum, Point};
+
+/// A trait for accessing data from a generic Point.
+pub trait PointTrait {
+    /// The coordinate type of this geometry
+    type T: CoordNum;
+
+    /// Access the n'th (0-based) element of the CoordinateTuple.
+    /// May panic if n >= DIMENSION.
+    /// See also [`nth()`](Self::nth).
+    fn nth_unchecked(&self, n: usize) -> Self::T;
+
+    /// Native dimension of the coordinate tuple
+    fn dim(&self) -> usize;
```

Comment:
I'm still split on whether this dimension should be physical or logical. Maybe we should have an enum like

```rs
pub enum Dimension {
    XY,
    XYZ,
    XYM,
    XYZM,
    Unknown(usize),
}
```

So in the common Simple Features dimension cases, we can transmit the logical dimension as well, while we still support other types of coordinates (maybe with time)?

When I think about the geometry representations I've implemented the traits on, all of them (except maybe GEOS?) have logical type information that is being lost through the trait.

Reply:
I think there is utility in having the explicit geometry dimension included like this. This is because of `XYZ` and `XYM` having the same dimension. 

In the case of `serde_esri` the [dimensions are tracked in the struct](https://github.com/JosiahParry/serde_esri/blob/28502b3a07280f09b92b9bc129e4f46ec6b23af1/src/geometry.rs#L139-L140)  but if passing to a geo method the distinction between `XYZ` and `XYM` may be quite important. 

For example, if one wanted to calculate distance based on dimension, one may conclude that a 3D distance (including altitude) may be appropriate if the dimension is `3` but that may not be the correct choice if the 3rd dimension is actually `M`.

Reply:
I think it was primarily @busstoptaktik who was advocating for physical, not logical type information here https://github.com/georust/geo/pull/1157#issuecomment-2007856305

Reply:
> Maybe we should have an enum like
>
> ```rust
> pub enum Dimension {
>    XY,
>    XYZ,
>    XYM,
>    XYZM,
>    Unknown(usize),
> }
>```

I think I like that.

If you know the logical layout, you can infer the physical. If you are dealing with some system where the layout is dynamic/unknowable at compile time, like what I understand @busstoptaktik to be talking about, you can still use output Unknown(usize).

We could make it slightly easier with something like:

```
impl Dimensions {
  fn len(&self) -> usize {
    switch self {
      XY => 2,
      XYZ | XYM => 3,
      XYZM => 4,
      Unknown(n) => n
    }
  }
}
```